clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7430,6 +7430,212 @@ static void handleSYCLIntelFPGAMaxConcurrencyAttr(Sema &S, Decl *D,
   S.AddSYCLIntelFPGAMaxConcurrencyAttr(D, A, E);
 }
 
+// Checks if an expression is a valid filter list for an add_ir_attributes_*
+// attribute. Returns true if an error occured.
+static bool checkAddIRAttributesFilterListExpr(Expr *FilterListArg, Sema *S,
+                                               const AttributeCommonInfo &CI) {
+  assert(isa<InitListExpr>(FilterListArg));
+  const auto *FilterListE = cast<InitListExpr>(FilterListArg);
+  for (const Expr *FilterElemE : FilterListE->inits())
+    if (!isa<StringLiteral>(FilterElemE))
+      return S->Diag(FilterElemE->getBeginLoc(),
+                     diag::err_sycl_add_ir_attribute_invalid_filter)
+             << CI;
+  return false;
+}
+
+// Returns true if a type is either an array of char or a pointer to char.
+static bool isAddIRAttributesValidStringType(QualType T) {
+  return (T->isArrayType() || T->isPointerType()) &&
+         T->getPointeeOrArrayElementType()->isCharType();
```

Comment:
Don't want want to check for `const char *` instead of `char *`?

Reply:
Good point. I have changed it so it checks the if the element type is `const` qualified. Sadly it means it cannot use `getPointeeOrArrayElementType` as it returns `clang::Type *`, but I think the solution looks relatively clean still.